working/0723-static-extensions/feature-specification-variant2.md

```diff
@@ -0,0 +1,347 @@
+parameters declared by an extension.
+
+An _explicitly resolved invocation_ of a constructor named `C.name` in a
+extension declaration _D_ named `E` with `s` type parameters and on-class
+`C` can be expressed as `E<S1 .. Ss>.C.name(args)`, 
+`E.C<U1 .. Uk>.name(args)`, or `E<S1 .. Ss>.C<U1 .. Uk>.name(args)` (and
+similarly for a constructor named `C` using `E<S1 .. Ss>.C(args)`, etc).
+
+A compile-time error occurs if the type arguments passed to `E` violate the
+declared bounds. A compile-time error occurs if no type arguments are
+passed to `E`, and type arguments `U1 .. Uk` are passed to `C`, but no list
+of actual type arguments for the type variables of `E` can be found such
+that the instantiated constructor return type of `E` with said type
+arguments is `C<U1 .. Uk>`.
+
+*Note that we must be able to choose the values of the type parameters
+`X1 .. Xs` such that the instantiated constructor return type is
+exactly `C<U1 .. Uk>`, it is not sufficient that it is a subtype thereof,
+or that it differs in any other way.*
+
+A compile-time error occurs if the invocation passes actual type arguments
+to both `E` and `C`, call them `S1 .. Ss` and `U1 .. Uk`, respectively,
+unless the instantiated constructor return type of _D_ with actual type
+arguments `S1 .. Ss` is `C<U1 .. Uk>`. In this type comparison, top types
+like `dynamic` and `Object?` are considered different, and no type
+normalization occurs. *In other words, the types must be equal, not
+just mutual subtypes.*
+
+*Note that explicitly resolved invocations of constructors declared in
+extensions are an exception in real code, usable in the case where a
+name clash prevents an implicitly resolved invocation. However, implicitly
+resolved invocations are specified in the rest of this section by reducing
+them to explicitly resolved ones.*
+
+A constructor invocation of the form `C<T1 .. Tm>.name(args)` is partially
+resolved by looking up a constructor named `C.name` in the class `C` and in
+every accessible extension with on-class `C`. A compile-time error
+occurs if no such constructor is found. Similarly, an invocation of the
+form `C<T1 ... Tm>(args)` uses a lookup for constructors named `C`.
+
+If a constructor in `C` with the requested name was found, the pre-feature
+static analysis and dynamic semantics apply. *That is, the class always
+wins.*
+
+Otherwise, the invocation is partially resolved to a set of candidate
+constructors found in extensions. Each of the candidates _kj_ is
+vetted as follows:
+
+Assume that _kj_ is a constructor declared by a extension _D_ named
+`E` with type parameters `X1 extends B1 .. Xs extends Bs` and on-type
+`C<S1 .. Sm>`.  Find actual values `U1 .. Us` for `X1 .. Xs` satisfying the
+bounds `B1 .. Bs`, such that `([U1/X1 .. Us/Xs]C<S1 .. Sm>) == C<T1 .. Tm>`.
+If this fails then remove _kj_ from the set of candidate constructors.
+Otherwise note that _kj_ uses actual type arguments `U1 .. Us`.
```

Comment:
I'd rather not remove candidates based on the `on` type.
It's great enough to have a constructor which can only be used for some instantiations, I don't want to also provide overloading, where
`List<int>.defaultFilled()` and `List<String>.defaultFilled()` refer to different constructors.

So, `type.id` always refers to precisely *one* declaration. If there is more than one static extension with that name in scope, you'll need to disambiguate. We don't do that based on the type.

(I know we failed thos for extension instance methods, where `List<int>` and `List<String>` can have different implementations. I think static members are trickier, and constructors should probably be treated as statics. That said, it's probably less error-prone to allow overloading of constructors based on context type than for instance members based on receiver type. The context is often more obvious than the current type of a promoted variable.)

Reply:
I created https://github.com/dart-lang/language/issues/4049 to discuss this.